compiler/rustc_passes/src/dead.rs

```diff
@@ -1177,8 +1183,16 @@ impl<'tcx> DeadVisitor<'tcx> {
             },
         };
 
+        let lint = if tcx.visibility(first_item.def_id).is_public()
+            && matches!(tcx.def_kind(first_item.def_id), DefKind::Struct)
+        {
+            UNCONSTRUCTIBLE_PUB_STRUCT
+        } else {
+            DEAD_CODE
+        };
```

Comment:
It seems very sketchy that this depends on `first_item`. Doesn't this suppress `dead_code` if a pub ADT comes first in the list?

Reply:
> It seems very sketchy that this depends on `first_item`. Doesn't this suppress `dead_code` if a pub ADT comes first in the list?

No, we will only have one pub struct there. Because this function will only be called by warn_dead_code for struct.